akka-actor-typed/src/main/scala/akka/actor/typed/internal/StashBufferImpl.scala

```diff
@@ -96,11 +101,36 @@ import akka.util.ConstantFun
 
   override def unstash(ctx: scaladsl.ActorContext[T], behavior: Behavior[T],
                        numberOfMessages: Int, wrap: T ⇒ T): Behavior[T] = {
-    val iter = new Iterator[T] {
-      override def hasNext: Boolean = StashBufferImpl.this.nonEmpty
-      override def next(): T = wrap(StashBufferImpl.this.dropHead())
-    }.take(numberOfMessages)
-    Behavior.interpretMessages[T](behavior, ctx, iter)
+    if (isEmpty)
+      behavior // optimization
+    else {
+      val iter = new Iterator[T] {
+        override def hasNext: Boolean = StashBufferImpl.this.nonEmpty
+        override def next(): T = wrap(StashBufferImpl.this.dropHead())
+      }.take(numberOfMessages)
+      interpretUnstashedMessages(behavior, ctx, iter)
+    }
+  }
+
+  private def interpretUnstashedMessages(behavior: Behavior[T], ctx: TypedActorContext[T], messages: Iterator[T]): Behavior[T] = {
+    @tailrec def interpretOne(b: Behavior[T]): Behavior[T] = {
+      val b2 = Behavior.start(b, ctx)
+      if (!Behavior.isAlive(b2) || !messages.hasNext) b2
+      else {
+        val nextB = try {
+          messages.next() match {
+            case sig: Signal ⇒ Behavior.interpretSignal(b2, ctx, sig)
+            case msg         ⇒ Behavior.interpretMessage(b2, ctx, msg)
+          }
+        } catch {
+          case NonFatal(e) ⇒ throw UnstashException(e, b2)
+        }
+
+        interpretOne(Behavior.canonicalize(nextB, b, ctx)) // recursive
```

Comment:
And the previous behavior should not be the started previous?

If unstashing into:
```
Behaviors.setup { _ => 

  Behaviors.receiveMessage { msg => 
    Behaviors.same
  }
}
```
I think this will make setup run for each unstashed message

Reply:
good catch, makes sense to use b2 here